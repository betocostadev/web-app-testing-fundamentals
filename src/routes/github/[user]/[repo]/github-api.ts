// Creating a simple fake API to fetch some repository data on Github.
// It needs a Private Github Access Token that you can generate in your Github Profile.
import type { paths } from '@octokit/openapi-types'

type OrgRepoResponse =
  paths['/repos/{owner}/{repo}']['get']['responses']['200']['content']['application/json']

export type Fetch = typeof fetch

// Fetch is passed here for test purposes. We are mocking a fetch with Vitest
export class GithubApi {
  constructor(
    private token: string | undefined,
    private fetch: Fetch
  ) {}

  async getRepository(user: string, repo: string) {
    const headers: HeadersInit = {
      'User-Agent': 'Qwik Workshop',
      'X-GitHub-Api-Version': '2022-11-28',
    }

    if (this.token) {
      headers['Authorization'] = 'Bearer ' + this.token
    }

    const response = await this.fetch(
      `https://api.github.com/repos/${user}/${repo}`,
      {
        headers,
      }
    )
    const repository = (await response.json()) as OrgRepoResponse

    return repository
  }
}
