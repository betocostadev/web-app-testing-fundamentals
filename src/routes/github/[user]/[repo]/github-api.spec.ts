import { describe, it } from 'vitest'
import { GithubApi } from './github-api'

describe('github-api', () => {
  // Understand that this isn't a unit test, since Github can go down and the test will fail
  describe('getRepository', () => {
    it('should return repository information', async ({ expect }) => {
      const api = new GithubApi(undefined)
      const response = await api.getRepository(
        'betocostadev',
        'web-app-testing-fundamentals'
      )
      expect(response).toMatchSnapshot()
    })
    it.todo('should timeout after x seconds with timeout response')
  })
})
