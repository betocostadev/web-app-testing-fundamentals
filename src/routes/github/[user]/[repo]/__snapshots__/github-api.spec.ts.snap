// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`github-api > getRepository > should return repository information 1`] = `
{
  "allow_forking": true,
  "archive_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/{archive_format}{/ref}",
  "archived": false,
  "assignees_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/assignees{/user}",
  "blobs_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/git/blobs{/sha}",
  "branches_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/branches{/branch}",
  "clone_url": "https://github.com/betocostadev/web-app-testing-fundamentals.git",
  "collaborators_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/collaborators{/collaborator}",
  "comments_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/comments{/number}",
  "commits_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/commits{/sha}",
  "compare_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/compare/{base}...{head}",
  "contents_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/contents/{+path}",
  "contributors_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/contributors",
  "created_at": "2024-10-16T19:18:28Z",
  "default_branch": "main",
  "deployments_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/deployments",
  "description": "Repository based on Web App Testing & Tools Fundamentals course by MHevery",
  "disabled": false,
  "downloads_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/downloads",
  "events_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/events",
  "fork": false,
  "forks": 0,
  "forks_count": 0,
  "forks_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/forks",
  "full_name": "betocostadev/web-app-testing-fundamentals",
  "git_commits_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/git/commits{/sha}",
  "git_refs_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/git/refs{/sha}",
  "git_tags_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/git/tags{/sha}",
  "git_url": "git://github.com/betocostadev/web-app-testing-fundamentals.git",
  "has_discussions": false,
  "has_downloads": true,
  "has_issues": true,
  "has_pages": false,
  "has_projects": true,
  "has_wiki": true,
  "homepage": null,
  "hooks_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/hooks",
  "html_url": "https://github.com/betocostadev/web-app-testing-fundamentals",
  "id": 873819498,
  "is_template": false,
  "issue_comment_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/issues/comments{/number}",
  "issue_events_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/issues/events{/number}",
  "issues_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/issues{/number}",
  "keys_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/keys{/key_id}",
  "labels_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/labels{/name}",
  "language": "TypeScript",
  "languages_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/languages",
  "license": {
    "key": "mit",
    "name": "MIT License",
    "node_id": "MDc6TGljZW5zZTEz",
    "spdx_id": "MIT",
    "url": "https://api.github.com/licenses/mit",
  },
  "merges_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/merges",
  "milestones_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/milestones{/number}",
  "mirror_url": null,
  "name": "web-app-testing-fundamentals",
  "network_count": 0,
  "node_id": "R_kgDONBVtag",
  "notifications_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/notifications{?since,all,participating}",
  "open_issues": 0,
  "open_issues_count": 0,
  "owner": {
    "avatar_url": "https://avatars.githubusercontent.com/u/9079952?v=4",
    "events_url": "https://api.github.com/users/betocostadev/events{/privacy}",
    "followers_url": "https://api.github.com/users/betocostadev/followers",
    "following_url": "https://api.github.com/users/betocostadev/following{/other_user}",
    "gists_url": "https://api.github.com/users/betocostadev/gists{/gist_id}",
    "gravatar_id": "",
    "html_url": "https://github.com/betocostadev",
    "id": 9079952,
    "login": "betocostadev",
    "node_id": "MDQ6VXNlcjkwNzk5NTI=",
    "organizations_url": "https://api.github.com/users/betocostadev/orgs",
    "received_events_url": "https://api.github.com/users/betocostadev/received_events",
    "repos_url": "https://api.github.com/users/betocostadev/repos",
    "site_admin": false,
    "starred_url": "https://api.github.com/users/betocostadev/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/betocostadev/subscriptions",
    "type": "User",
    "url": "https://api.github.com/users/betocostadev",
    "user_view_type": "public",
  },
  "private": false,
  "pulls_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/pulls{/number}",
  "pushed_at": "2024-10-21T19:36:04Z",
  "releases_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/releases{/id}",
  "size": 15412,
  "ssh_url": "git@github.com:betocostadev/web-app-testing-fundamentals.git",
  "stargazers_count": 0,
  "stargazers_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/stargazers",
  "statuses_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/statuses/{sha}",
  "subscribers_count": 1,
  "subscribers_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/subscribers",
  "subscription_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/subscription",
  "svn_url": "https://github.com/betocostadev/web-app-testing-fundamentals",
  "tags_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/tags",
  "teams_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/teams",
  "temp_clone_token": null,
  "topics": [],
  "trees_url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals/git/trees{/sha}",
  "updated_at": "2024-10-21T19:36:08Z",
  "url": "https://api.github.com/repos/betocostadev/web-app-testing-fundamentals",
  "visibility": "public",
  "watchers": 0,
  "watchers_count": 0,
  "web_commit_signoff_required": false,
}
`;
